src/framework/MOM_intrinsic_functions.F90

```diff
@@ -25,4 +28,127 @@ function invcosh(x)
+  real :: root !< The real cube root of x in arbitrary units [A]
+
+  real :: asx ! The absolute value of x rescaled by an integer power of 8 to put it into
+              ! the range from 0.125 < asx <= 1.0, in ambiguous units cubed [B3]
+  real :: root_asx ! The cube root of asx [B]
+  real :: num ! The numerator of an expression for the evolving estimate of the cube root of asx
+              ! in arbitrary units that can grow or shrink with each iteration [B C]
+  real :: den ! The denominator of an expression for the evolving estimate of the cube root of asx
+              ! in arbitrary units that can grow or shrink with each iteration [C]
+  real :: num_prev ! The numerator of an expression for the previous iteration of the evolving estimate
+              ! of the cube root of asx in arbitrary units that can grow or shrink with each iteration [B D]
+  real :: den_prev ! The denominator of an expression for the previous iteration of the evolving estimate of
+              ! the cube root of asx in arbitrary units that can grow or shrink with each iteration [D]
+  integer :: ex_3 ! One third of the exponent part of x, used to rescale x to get a.
+  integer :: itt
+
+  if ((x >= 0.0) .eqv. (x <= 0.0)) then
+    ! Return 0 for an input of 0, or NaN for a NaN input.
+    root = x
+  else
+    ex_3 = ceiling(exponent(x) / 3.)
+    ! Here asx is in the range of 0.125 <= asx < 1.0
+    asx = scale(abs(x), -3*ex_3)
+
+    ! This first estimate is one iteration of Newton's method with a starting guess of 1.  It is
+    ! always an over-estimate of the true solution, but it is a good approximation for asx near 1.
+    num = (2.0 + asx)
+    den = 3.0
+    ! Iteratively determine Root = asx**1/3 using Newton's method, noting that in this case Newton's
+    ! method converges monotonically from above and needs no bounding.  For the range of asx from
+    ! 0.125 to 1.0 with the first guess used above, 6 iterations suffice to converge to roundoff.
+    do itt=1,9
+      ! Newton's method iterates estimates as Root = Root - (Root**3 - asx) / (3.0 * Root**2), or
+      ! equivalently as Root = (2.0*Root**2 + asx) / (3.0 * Root**2).
+      ! Keeping the estimates in a fractional form Root = num / den allows this calculation with
+      ! fewer (or no) real divisions during the iterations before doing a single real division
+      ! at the end, and it is therefore more computationally efficient.
+
+      ! Because successive estimates of the numerator and denominator tend to be the cube of their
+      ! predecessors, the numerator and denominator need to be rescaled by division when they get
+      ! too large or small to avoid overflow or underflow in the convergence test below.  Were
+      ! this being done for 32 bit reals, the values to compare with would be about 1.0e9 and 1.0e-9.
+      if ((den > 1.0e75) .or. (den < 1.0e-75)) then
+        num = num / den ; den = 1.0
+      endif
```

Comment:
This block could be written as follows:
```
real, parameter :: den_min = 2.**(minexponent(1.) / 4)
real, parameter :: den_max = 2.**(maxexponent(1.) / 4)

! ...

if (den > den_max .or. den < den_min) then
  num = scale(num, -exponent(den))
  den = scale(den, -exponent(den))
endif
````
This would avoid an extra derivative, and also write it in a largely precision-independent form.

I'd also suggest moving it to the end of the loop, since the initial value will never require scaling, and one of the `exit` statements may skip the if-block.
(Depending on the circumstances, it may even be cheaper to do this every iteration, but I'm already overthinking this problem.)

Reply:
Thank you for these suggestion; I have now implemented them in the revised PR.